From aaea0fc13842636103f6b7a5143d6b287e23faf9 Mon Sep 17 00:00:00 2001
From: Jannis Harder <jix@jixco.de>
Date: Fri, 27 Jul 2012 23:19:53 +0200
Subject: [PATCH 05/79] Add segments to label addresses

Generate addresses with a segment part for labels inside a section
with a defined vseg
---
 modules/objfmts/bin/bin-objfmt.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/modules/objfmts/bin/bin-objfmt.c b/modules/objfmts/bin/bin-objfmt.c
index 7748f1e4..5286c7ee 100644
--- a/modules/objfmts/bin/bin-objfmt.c
+++ b/modules/objfmts/bin/bin-objfmt.c
@@ -533,8 +533,20 @@ bin_objfmt_expr_xform(/*@returned@*/ /*@only@*/ yasm_expr *e,
             bsd = yasm_section_get_data(sect, &bin_section_data_cb);
             assert(bsd != NULL);
             yasm_intnum_calc(dist, YASM_EXPR_ADD, bsd->ivstart);
-            e->terms[i].type = YASM_EXPR_INT;
-            e->terms[i].data.intn = dist;
+            if (bsd->ivseg != NULL) {
+                yasm_expr *sube;
+                yasm_intnum *seg;
+
+                seg = yasm_intnum_copy(bsd->ivseg);
+                sube = yasm_expr_create(YASM_EXPR_SEGOFF,
+                                        yasm_expr_int(seg),
+                                        yasm_expr_int(dist), e->line);
+                e->terms[i].type = YASM_EXPR_EXPR;
+                e->terms[i].data.expn = sube;
+            } else {
+                e->terms[i].type = YASM_EXPR_INT;
+                e->terms[i].data.intn = dist;
+            }
         }
 
         /* Transform our special symrecs into the appropriate value */
-- 
2.49.0

